### PR TITLE
(MINOR) Add support for manifest removal

### PR DESCRIPTION
**Pull request by mauricefisher64**
## Changes in this pull request
New internal API to allow for the removal of a manifest.  This will be needed for future use cases 

## Checklist
- [x] This PR represents a single feature, fix, or change.
- [x] All applicable changes have been documented.
- [x] Any `TO DO` items (or similar) have been entered as GitHub issues and the link to that issue has been included in a comment.

### CODE REVIEW

File: sdk/src/asset_handlers/bmff_io.rs [hunk 1]
```diff
@@ -1066,6 +1066,135 @@ impl AssetIO for BmffIO {
...
+        };
+
+        // write content before ContentProvenanceBox
+        input.seek(SeekFrom::Start(0))?;
+        let mut b = vec![0u8; start];
+        input.read_exact(&mut b)?;
+        temp_file.write_all(&b)?;
+
+        // calc offset adjustments
+        // value will be negative sibce the box is truncated
```
Comment (scouten-adobe) [inline sdk/src/asset_handlers/bmff_io.rs:1133]: Typo: sibce -> since

File: sdk/src/asset_handlers/bmff_io.rs [hunk 2]
```diff
@@ -1066,6 +1066,135 @@ impl AssetIO for BmffIO {
...
+                .map_err(|_err| Error::BadParam("value out of range".to_string()))?;
+
+            (start, end)
+        } else {
+            return Err(Error::BadParam("value out of range".to_string()));
+        };
+
+        // write content before ContentProvenanceBox
+        input.seek(SeekFrom::Start(0))?;
+        let mut b = vec![0u8; start];
```
Comment (scouten-adobe) [inline sdk/src/asset_handlers/bmff_io.rs:1128]: This will cause problems for especially large files, right? (I'm thinking about video.)

File: sdk/src/asset_handlers/bmff_io.rs [hunk 3]
```diff
@@ -1257,4 +1386,27 @@ pub mod tests {
...
+        let mut success = false;
+        if let Ok(temp_dir) = tempdir() {
+            let output = temp_dir_path(&temp_dir, "mp4_test.mp4");
+
+            if let Ok(_size) = std::fs::copy(&source, &output) {
+                let bmff = BmffIO::new("mp4");
+
+                if let Ok(()) = bmff.remove_cai_store(&output) {
+                    match bmff.read_cai_store(&output) {
+                        Ok(_) => success = false,
```
Comment (scouten-adobe) [inline sdk/src/asset_handlers/bmff_io.rs:1403]: Assigning to `success` and matching makes it a bit hard to follow the overall control flow.

Perhaps you could use `if let ... else` and have more specific assertions at the point where each failure case is detected?

File: sdk/src/asset_handlers/c2pa_io.rs [hunk 4]
```diff
@@ -60,6 +60,10 @@ impl AssetIO for C2paIO {
 
         Ok(vec![hop])
     }
+
+    fn remove_cai_store(&self, _asset_path: &Path) -> Result<()> {
+        Ok(())
```
Comment (scouten-adobe) [inline sdk/src/asset_handlers/c2pa_io.rs:65]: Seems like this should be some kind of error signaling a non-sensical request instead of `Ok(())`.

File: sdk/src/asset_handlers/png_io.rs [hunk 5]
```diff
@@ -384,6 +384,44 @@ impl AssetIO for PngIO {
...
+                chunk crc (4 bytes in crc in format defined in PNG spec)
+        */
+
+        // erase existing
+        let empty_buf = Vec::new();
+        let mut iter = ps.into_iter();
+        if let Some(existing_cai) = iter.find(|pcp| pcp.name == CAI_CHUNK) {
+            // replace existing CAI
+            let start = usize::value_from(existing_cai.start)
+                .map_err(|_err| Error::BadParam("value out of range".to_string()))?; // get beginning of chunk which starts 4 bytes before label
```
Comment (scouten-adobe) [inline sdk/src/asset_handlers/png_io.rs:411]: I'm not sure `BadParam` (typically error on the part of the caller) makes sense here. This is really saying the existing file is has faulty data in it, right?

File: sdk/src/asset_handlers/bmff_io.rs (hunk 2 above)
Comment (mauricefisher64) [reply sdk/src/asset_handlers/bmff_io.rs:1128]: No, the C2PA box is right after the ftyp box.  The ftyp is box first and is usually about 25-30 bytes so we only have a few bytes to copy before we hit the C2PA box.  After that notice we do chunk the large portion of the file.

File: sdk/src/asset_handlers/bmff_io.rs (hunk 1 above)
Comment (mauricefisher64) [reply sdk/src/asset_handlers/bmff_io.rs]: Fixed

File: sdk/src/asset_handlers/png_io.rs (hunk 5 above)
Comment (mauricefisher64) [reply sdk/src/asset_handlers/png_io.rs:411]: Or the start cannot fit into a usize.

File: sdk/src/asset_handlers/c2pa_io.rs (hunk 4 above)
Comment (mauricefisher64) [reply sdk/src/asset_handlers/c2pa_io.rs:65]: Yeah, I am not sure about this one.  No one will ever call this so success or failure does not matter at this time.

File: sdk/src/asset_handlers/bmff_io.rs [hunk 6]
```diff
@@ -1066,6 +1066,135 @@ impl AssetIO for BmffIO {
         let vec: Vec<HashObjectPositions> = Vec::new();
         Ok(vec)
     }
+
+    fn remove_cai_store(&self, asset_path: &Path) -> Result<()> {
```
Comment (gpeacock) [inline sdk/src/asset_handlers/bmff_io.rs:1070]: This should really be a READ+SEEK trait but I realize there's temp_files inside so there's no point. I guess eventually we need trait abstraction for a temp file? Not to worry about now, but I want us to get there evenutally.

File: sdk/src/asset_handlers/bmff_io.rs [hunk 7]
```diff
@@ -1066,6 +1066,135 @@ impl AssetIO for BmffIO {
...
+            };
+
+        let mut temp_file = Builder::new()
+            .prefix("c2pa_temp")
+            .rand_bytes(5)
+            .tempfile()?;
+
+        let (start, end) = if let Some(c2pa_length) = c2pa_length {
+            let start = usize::value_from(c2pa_start)
+                .map_err(|_err| Error::BadParam("value out of range".to_string()))?; // get beginning of chunk which starts 4 bytes before label
```
Comment (gpeacock) [inline sdk/src/asset_handlers/bmff_io.rs:1116]: This is a pretty obscure error if this ever comes up. It isn't really a bad parameter, more like invalid bmff data?

File: sdk/src/asset_handlers/png_io.rs (hunk 5 above)
Comment (gpeacock) [reply sdk/src/asset_handlers/png_io.rs:411]: Hey, same comment I made above. I think we should identify these as invalid bmff content.

**Review by gpeacock — CHANGES_REQUESTED** [review]
The exp-test1.png fixture is way larger than all other image we have in the sdk combined. We should try to shrink it down (that 10mb crate limit we have to deal with).

File: sdk/src/asset_handlers/c2pa_io.rs (hunk 4 above)
Comment (mauricefisher64) [reply sdk/src/asset_handlers/c2pa_io.rs:65]: I don't want to return an error for a no op

File: sdk/src/asset_handlers/bmff_io.rs [hunk 8]
```diff
@@ -1257,4 +1386,25 @@ pub mod tests {
         }
         assert!(success)
     }
+
+    #[test]
+    fn test_remove_c2pa() {
+        let source = fixture_path("video1.mp4");
+
+        if let Ok(temp_dir) = tempdir() {
```
Comment (scouten-adobe) [inline sdk/src/asset_handlers/bmff_io.rs:1394]: Since you're in test code, why not use `.unwrap()` rather than the nested `if let`s?

File: sdk/src/asset_handlers/c2pa_io.rs (hunk 4 above)
Comment (scouten-adobe) [reply sdk/src/asset_handlers/c2pa_io.rs:65]: Agree with leaving this as `Ok(())`, but can we at least log that this was a non-sensical request?

File: sdk/src/asset_handlers/jpeg_io.rs [hunk 9]
```diff
@@ -431,4 +450,25 @@ pub mod tests {
         let result = extract_xmp(&seg);
         assert_eq!(result, None);
     }
+
+    #[test]
+    fn test_remove_c2pa() {
+        let source = crate::utils::test::fixture_path("CA.jpg");
+
+        if let Ok(temp_dir) = tempfile::tempdir() {
```
Comment (scouten-adobe) [inline sdk/src/asset_handlers/jpeg_io.rs:458]: Same comment about `if let` vs `.unwrap()`.

File: sdk/src/asset_handlers/png_io.rs [hunk 10]
```diff
@@ -433,4 +469,25 @@ pub mod tests {
             }
         }
     }
+
+    #[test]
+    fn test_remove_c2pa() {
+        let source = crate::utils::test::fixture_path("exp-test1.png");
+
+        if let Ok(temp_dir) = tempfile::tempdir() {
```
Comment (scouten-adobe) [inline sdk/src/asset_handlers/png_io.rs:477]: Here too …

File: sdk/src/asset_handlers/bmff_io.rs (hunk 8 above)
Comment (mauricefisher64) [reply sdk/src/asset_handlers/bmff_io.rs:1394]: Some intermediate version of Clippy made me do it.  If necessary I can see if it is still an issue.

**Review by gpeacock — APPROVED** [review]
